server/src/main/java/org/opensearch/cluster/metadata/Metadata.java

```diff
@@ -1329,6 +1336,25 @@ public Builder removeDataStream(String name) {
             return this;
         }
 
+        public Builder queryGroups(final Set<QueryGroup> queryGroups) {
+            this.customs.put(QueryGroupMetadata.TYPE, new QueryGroupMetadata(queryGroups));
+            return this;
+        }
+
+        public Builder put(final QueryGroup queryGroup) {
+            Objects.requireNonNull(queryGroup, "queryGroup should not be null");
+            Set<QueryGroup> existing = getQueryGroups();
+            existing.add(queryGroup);
+            return queryGroups(existing);
+        }
+
+        private Set<QueryGroup> getQueryGroups() {
+            return Optional.ofNullable(this.customs.get(QueryGroupMetadata.TYPE))
+                .map(o -> (QueryGroupMetadata) o)
+                .map(QueryGroupMetadata::queryGroups)
+                .orElse(Collections.emptySet());
+        }
+
```

Comment:
public Builder put(final QueryGroup queryGroup) {
            Objects.requireNonNull(queryGroup, "queryGroup should not be null");
            Set<QueryGroup> existing = getQueryGroups();
            HashSet<QueryGroup> updatedGroups = new HashSet<>(existing);
            updatedGroups.add(queryGroup);
            return queryGroups(updatedGroups);
        }

        public Builder remove(final QueryGroup queryGroup) {
            Objects.requireNonNull(queryGroup, "queryGroup should not be null");
            Set<QueryGroup> existing = getQueryGroups();
            HashSet<QueryGroup> updatedGroups = new HashSet<>(existing);
            updatedGroups.remove(queryGroup);
            return queryGroups(updatedGroups);
        }

Reply:
Yes, we need to send the new copy instead of modifying the existing one.